proposals/manual-bundling.md

```diff
@@ -0,0 +1,150 @@
+
+- Package managers ([npm] / [yarn])
+- Module bundlers ([webpack] / [browserify] / [rollup])
+- IDEs (Atom ([autocomplete-modules], [js-hyperclick]))
+- Linters ([eslint-plugin-import])
+- Miscellaneous tooling ([depcheck])
+
+
+## Current situation
+
+Let's say I would like to build a Cordova app for Android using the camera plugin.
+
+Steps:
+
+1. Create a Cordova project using `cordova create`.
+2. Remove the legacy hooks directory and obsolete .npmignore.
+3. Add the desired Android platform and camera plugin.
+
+   Now this is where things get weird.
+
+   A directory `/platforms/android/platform_www` is created.
+
+   Inside this folder is a file `cordova.js`, which contains a UMD bundle with the following modules:
+
+   - `'cordova'` → https://github.com/apache/cordova-js/blob/master/src/cordova.js
+   - `'cordova/exec'` → https://github.com/apache/cordova-android/blob/master/cordova-js-src/exec.js
+   - `'cordova/platform'` → https://github.com/apache/cordova-android/blob/master/cordova-js-src/platform.js
+   - `'cordova/android/*'` → https://github.com/apache/cordova-android/tree/master/cordova-js-src/android
+   - `'cordova/plugin/android/*'` → https://github.com/apache/cordova-android/tree/master/cordova-js-src/plugin/android
+   - `'cordova/*'` → https://github.com/apache/cordova-js/tree/master/src/common (except for exec and platform, not sure if they're the `_b` postfixed ones, but it doesn't really matter)
+
+   In addition there is a `cordova_plugins.js`, which maps the plugins to a file location.
+
+   Also there is a `plugins/cordova-plugin-camera/www/*`, which contains the files from https://github.com/apache/cordova-plugin-camera/tree/master/www, but with a `.` as a path separator, whereas in any other situation this would be a `/`.
+
+   However, if I take a look at what the `cordova` package on npm is, it's the CLI, which is totally unrelated to the UMD bundle.
+   
+   In addition to `/platforms`, there's also a `/plugins` directory. This contains the following:
+   
+   - Copies of the plugins that already exist in `node_modules`, but with the dotfiles missing.
+   - `android.json` — This appears to duplicate some info from `package.json` and `config.xml`, but in another format.
+   - `fetch.json` — This appears to duplicate some info from `package.json`, but in yet another format.
+
+
+## Proposal
+
+As a modern JavaScript developer, I would like to use *<this year's popular tool>* to bundle my dependencies and code.
+
+- Get rid of `cordova.js`.
+- Get rid of `www/`.
+- Get rid of `plugins/`.
+
+### 1. Load plugins from the `node_modules` directory, not `plugins`
+
+Cordova already downloads packages to `node_modules` using `cordova-fetch`. There's no need to maintain a copy of them in the `plugins` directory.
+
+This also allows plugin creators to use npm dependencies for both hooks and browser code the way this would be done for other packages.
+
+### 2. Let a developer require Cordova using its own build tools.
```

Comment:
> So if i'm understanding this correctly, you are suggesting we use html imports to do plugin loading.

I'm not suggesting to use html imports, but I'm suggesting to require a build tool. These tools are already specialized at handling node modules, so Cordova doesn't have to handle this differently. Anywhere where I write "`import`", it might just as well say "`require`".

> Few concerns:

> - I think this would be a tough sell for plugins. Pretty much every plugin would have to change.

One implementation of such a build hook could mimic the old behaviour for a while.

- Copy the contents of `wwww` into the platform www directory.
- Use browserify (or webpack) to detect all Cordova plugins and the Cordova platform directory, bundle those and write this into `cordova.js` in the platform www directory.
- For backwards compatibility, this could map `cordova` imports to `cordova-js` and `cordova/*` to `cordova-js/*`.

This could be included by default in `cordova-app-hello-world`.

I would personally prefer a hook that would bundle and minify everything, including my own code, my own dependnecies and Cordova code. This means I would have to write my own hook for this.


> - Our goal is to mimic web apis. Camera is a bad example because it isn't based on a standard. But a lot of the current work on our plugin audit is about deprecating non standard apis (We are keeping camera don't worry) and updating apis of plugins where the standard has changed since we implemented it originally.

Mimicking (or polyfilling) web APIs doesn't require any special handling from Cordova. Just assign values to the `window` object.


> - imports aren't supported everywhere i think. Would that force users to use a build tool?

This would force users to include a build hook. This is likely to be based on a build tool.